File: 1_Two Sum.cpp
#include <map>
#include <vector>

class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target) {
	    std::unordered_map<int, int> map;
	    std::unordered_map<int, int>::iterator it;
	    for (int i = 0; i < numbers.size(); ++i) {
		    if ((it = map.find(target - numbers[i])) != map.end())
			    return std::vector<int> {it->second + 1, i + 1};
		    else
			    map.insert(std::make_pair(numbers[i], i));
	    }		    
    }
};

File: 1_Two Sum.py
class Solution:
    # @return a tuple, (index1, index2)
    def twoSum(self, num, target):
        enum = list(enumerate(num))
        enum.sort(key = lambda x: x[1])
        left = 0
        right = len(num) - 1
        while True:  #optimization
            while enum[left][1] + enum[right][1] > target:
                right -= 1
            if enum[left][1] + enum[right][1] == target:
                return sorted((enum[left][0] + 1, enum[right][0] + 1))
            left += 1
